### DISCUSSION TITLE
some questions around the encoding

### DISCUSSION THREAD

**Discussion by parsisoftware** [pyg-team/pytorch-frame#252] [Q&A]
Hello, I am trying to use pytorch-frame in pytorch geometric. I have some questions.
1) Can I use column-wise encoding for each column, not stype-wise?

2) I have used the following way as it was provided in the examples:
text_encoder = TextToEmbedding(model=args.model, pooling=args.pooling,
                                   device=device)
text_embedder_cfg = TextEmbedderConfig(text_embedder=text_encoder,
                                   batch_size=5)

But it takes a long time  to embed texts: 
Embedding texts in mini-batch:  95%|█████████▍| 1848/1949 [1:37:36<05:47,  3.44s/it]
Am I doing wrong?

-3) How do I know which possible encoders can be used for each stype?

4) I want to create graph before I go through defining a neural model. I saw in the examples that the encoding of each stype is assigned in the model. I want to present the graph with the attribute values (according to the encoding) before defining the models. 
Is it possible?

5) my nodes and edges have some attributes of different types. Is it possible to use this library even in situation where there is no need to consider attributes? (I mean one code for both situations)

6) what the following error is?

finite_mask = np.isfinite(flattened)
TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

I am new in this field. I am waiting for your helps.
Thanks.

**Comment by yiweny** [comment]
Hi @parsisoftware 
1. We are planning on implementing column-wise encoding and putting it in 0.2.0 release. See #232 
2. It's expected to take a long time if your text is huge. You can try increasing the batch size?
3. You can see the full list of encoders [here](https://pytorch-frame.readthedocs.io/en/latest/modules/nn.html#torch-frame-nn-encoder). 
4. I am a bit confused by your question. When are you looking to define to model then?
5. deferring to @rusty1s for this
6. Can you give more information, e.g. print out the values of the `flattened`? Thanks~

**Reply by weihua916** [reply]
Just to add for 2. Transformer is generally very expensive. You can implement lighter text embedders, like Glove averaging.

**Comment by parsisoftware** [comment]
Thank you for your response.
When the column-wise encoding will be ready? I have to use it soon.

Here is the value  of flattened:
('fluttened', array([  2,   2,   2, ..., 610, 610, 610], dtype=int64))
('fluttened', array([ 60756,  60756,  60756, ...,   3265,   3265, 168248], dtype=int64))
('fluttened', array(['funny', 'Highly quotable', 'will ferrell', ..., 'gun fu',
       'heroic bloodshed', 'Heroic Bloodshed'], dtype='<U85'))

I have tried to encode the feature in the movie example in the following link:
https://pytorch-geometric.readthedocs.io/en/latest/tutorial/load_csv.html

**Reply by weihua916** [reply]
> When the column-wise encoding will be ready? I have to use it soon.

It will be ready by the end of this year.

> ('fluttened', array(['funny', 'Highly quotable', 'will ferrell', ..., 'gun fu',
'heroic bloodshed', 'Heroic Bloodshed'], dtype='<U85'))

Looks like you've specified this column to be numerical...

**Reply by parsisoftware** [reply]
I have defined the column types manually in this way:

for f in path.glob('*.csv'):
    filename= Path(f).stem
    df=pd.read_csv(f)
    df_dict[filename]=df

    if filename == "links":
        col_to_stype_dict[filename]={}
        col_to_stype_dict[filename]["movieId"]= torch_frame.numerical
        col_to_stype_dict[filename]["imdbId"] = torch_frame.numerical
        col_to_stype_dict[filename]["tmdbId"] = torch_frame.numerical

    if filename == "movies":
        col_to_stype_dict[filename]={}
        col_to_stype_dict[filename]["movieId"] = torch_frame.numerical
        col_to_stype_dict[filename]["title"] = torch_frame.text_embedded
        col_to_stype_dict[filename]["genres"]= torch_frame.categorical

    if filename == "ratings":
        col_to_stype_dict[filename]= {}
        col_to_stype_dict[filename]["userId"] = torch_frame.numerical
        col_to_stype_dict[filename]["movieId"] = torch_frame.numerical
        col_to_stype_dict [filename]["rating"] = torch_frame.numerical
        col_to_stype_dict ["timestamp"] = torch_frame.numerical

    if filename == "tags":
        col_to_stype_dict[filename] = {}
        col_to_stype_dict[filename]["userId"] = torch_frame.numerical
        col_to_stype_dict[filename]["movieId"] = torch_frame.numerical
        col_to_stype_dict[filename]["tag"] = torch_frame.numerical
        col_to_stype_dict["timestamp"] = torch_frame.numerical

**Reply by weihua916** [reply]
I see. I think `col_to_stype_dict[filename]["tag"] = torch_frame.numerical` is the problematic part.

**Reply by parsisoftware** [reply]
you're right.
Thank you
